Implement directory synchronization

senofsky opened the pull request:
This PR adds the `-copy-in-before` and `-copy-out-after` feature. Passing tests are included in this PR.

A few concerns:
I had to add: `PATH_TO_IMAGE_STORE = f'{os.environ["HOME"]}/.local/share/transient/'` in `steps.py` because I needed a way to access the image path. I'm not sure if this is the cleanest way.
This is a nitpick but there doesn't seem to be consistent use of single and double quotes for python strings. I'm hoping this is taken care of when the black formatter is added.

senofsky committed:
added copy-in-before and copy-out-after feature to transient.py

senofsky committed:
added tests for copy-in-before and copy-out-after

senofsky committed:
isolated test artifacts to 'test/artifacts' directory for easier comprehension and cleanup

ALSchwalm commented on test/features/steps/steps.py:
```diff
@@ -145,3 +174,33 @@ def step_impl(context, name):
+    mirrored_guest_directory = os.path.join('artifacts', 'guest')
+    os.makedirs(mirrored_guest_directory, exist_ok=True)
+
+    image_path = os.path.join(PATH_TO_IMAGE_STORE,'test-vm-0-generic_alpine38_v3.0.2')
```
I'm pretty sure this path isn't correct, as we now encode the elements of the frontend path in a more complex way

ALSchwalm commented on test/features/steps/steps.py:
```diff
@@ -1,8 +1,11 @@
 from behave import *
 from hamcrest import *
 
+PATH_TO_IMAGE_STORE = f'{os.environ["HOME"]}/.local/share/transient/'
```
I don't understand why we would need this. If we want to interact with something in the frontend/backend in a test. We should almost certainly be setting the frontend/backend image directories to be under the test dir.

ALSchwalm commented on test/features/steps/steps.py:
```diff
@@ -30,6 +33,9 @@ def wait_on_vm(context, timeout=VM_WAIT_TIME):
     context.stderr = context.handle.stderr.read().decode('utf-8')
 
+def get_deepest_directory_name(directory):
+    return directory.split('/')[-2]
```
https://www.memesmonkey.com/images/memesmonkey/94/94745eb5d13997420e6089d6486d8334.jpeg

But seriously, some kind of comment about what this is doing/for would be good? The name does not make it clear to me.

senofsky replied:
This function was deleted after refactoring.

ALSchwalm commented on transient/transient.py:
```diff
@@ -46,6 +46,70 @@ def __needs_ssh_console(self) -> bool:
+        try:
+            vm_absolute_path, host_directory = path_mapping.split(':')
+        except ValueError:
+            raise RuntimeError(f'ERROR: invalid file mapping: {path_mapping}' +
```
We don't to this 'ERROR' prefix anywhere else, please avoid it. Also, having a newline in the text of a runtime error seems odd to me.

senofsky committed:
add caveat in README about -copy-in-before and -copy-out-after

ALSchwalm commented on test/features/copy-in-copy-out.feature:
```diff
@@ -0,0 +1,32 @@
+    - copy the guest file or directory to the host directory after stopping the VM
+
+  @skip-in-ci
+  Scenario: Copy in directory before starting VM
+     Given a transient vm
+       And a name "test-vm"
+       And a disk image "generic/alpine38:v3.0.2"
+       And a host directory: "artifacts/host/copy-in-test/"
+       And a guest directory: "/home/vagrant/"
+       And the host directory is copied to the guest directory before starting
+       And the vm is prepare-only
+      When the transient command is run
+      Then the return code is 0
+       And the host directory exists in the guest directory
+
+  @skip-in-ci
+  Scenario: Copy out directory after stopping VM
+     Given a transient vm
+       And a name "test-vm"
+       And a disk image "generic/alpine38:v3.0.2"
+       And a host directory: "artifacts/host/copy-out-test/"
+       And a guest directory: "/home/vagrant/"
+       And the guest directory is copied to the host directory after stopping
+       And a ssh command "echo 'dummy command'"
+      When the vm runs to completion
+      Then the return code is 0
+       And the guest directory exists in the host directory
```
I think I'd like to test this a bit differently. Instead of running virt-copy-in and virt-copy-out in the tests, we should probably just using `-copy-in-before` to add a file like `my-test-file` to the guest home directory, the run a guest with ssh command `ls` and assert that `my-test-file` is in stdout.

Similarly, to test `-copy-out-after` we can run an ssh command that does a `touch my-test-file` and then copy that out and assert that the file exists on the host after.

ALSchwalm commented on test/features/copy-in-copy-out.feature:
```diff
@@ -0,0 +1,32 @@
+    - copy the host file or directory to the guest directory before starting the VM
+    - copy the guest file or directory to the host directory after stopping the VM
+
+  @skip-in-ci
```
You might just be doing this for testing (if so please ignore), but before we merge we really need to make this only not run the tests in the docker part of CI, rather than _all_ CI. We should just be able to make a new tag and makefile target to do that.

ALSchwalm reviewed (changes requested):
Nice work! I just had a few comments/suggestions.

senofsky committed:
added libguestfs-tools to package dependencies, changed which tests
docker invokes, and changed kernel permissions to allow read from
non-root users so libguestfs-tools can work

ALSchwalm commented on transient/transient.py:
```diff
@@ -46,6 +46,72 @@ def __needs_ssh_console(self) -> bool:
+            raise RuntimeError(f'Absolute path for guest required: {vm_absolute_directory}')
+
+        env = os.environ
+        env['LIBGUESTFS_BACKEND'] = 'direct'
```
Please add a comment explaining that this is to avoid a libvirt dependency from libguestfs

ALSchwalm commented on transient/transient.py:
```diff
@@ -46,6 +54,74 @@ def __needs_ssh_console(self) -> bool:
                 self.config.ssh_command is not None)
 
+    def __get_required_environment(self) -> Environ:
+        """Configures the environment to remove the libvirt dependency from libguestfs-tools"""
+        env = os.environ
+        env['LIBGUESTFS_BACKEND'] = 'direct'
```
Haha, nice. I was thinking that should just be in a common function. Unfortunate about the Environ stuff. I'm increasingly not a fan of mypy

ALSchwalm commented:
Looks good to me. Merging.

One minor note for the future is that it's generally better to have shorter first lines of the commit message, then the longer description after that.
